base/iddict.jl

```diff
@@ -83,18 +84,31 @@ function sizehint!(d::IdDict, newsz)
     rehash!(d, newsz)
 end
 
-function setindex!(d::IdDict{K,V}, @nospecialize(val), @nospecialize(key)) where {K, V}
+function ht_keyindex!(d::IdDict{K,V}, @nospecialize(key)) where {K, V}
     !isa(key, K) && throw(KeyTypeError(K, key))
-    if !(val isa V) # avoid a dynamic call
-        val = convert(V, val)::V
-    end
+    keyindex = ccall(:jl_eqtable_keyindex, Cssize_t, (Any, Any), d, key)
+    # keyindex - where a key is stored, or -pos if the key was not present and was inserted at pos
+
+    return abs(keyindex), keyindex < 0
+end
+
+function _setindex!(d::IdDict{K,V}, val::V, keyindex::Int, inserted::Bool) where {K, V}
+    @inbounds d.ht[keyindex+1] = val
+    d.count += inserted
+
     if d.ndel >= ((3*length(d.ht))>>2)
         rehash!(d, max((length(d.ht)%UInt)>>1, 32))
         d.ndel = 0
     end
-    inserted = RefValue{Cint}(0)
-    d.ht = ccall(:jl_eqtable_put, Memory{Any}, (Any, Any, Any, Ptr{Cint}), d.ht, key, val, inserted)
-    d.count += inserted[]
+    return nothing
+end
+
+function setindex!(d::IdDict{K,V}, @nospecialize(val), @nospecialize(key)) where {K, V}
+    keyindex, inserted = ht_keyindex!(d, key)
+    if !(val isa V) # avoid a dynamic call
+        val = convert(V, val)::V
+    end
+    _setindex!(d, val, keyindex, inserted)
```

Comment:
Following `dict.jl` API, the `@inbounds` should be here rather than inside `_setindex!` (which gets marked with `@propagate_inbounds`